sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -77,6 +77,14 @@ case class Project(projectList: Seq[NamedExpression], child: LogicalPlan)
 
   override lazy val validConstraints: ExpressionSet =
     getAllValidConstraints(projectList)
+
+  override def metadataOutput: Seq[Attribute] =
+    child.metadataOutput.filter(_.isHiddenCol) ++
```

Comment:
This is needed to propagate hidden columns through nested Projections (eg. multiple natural joins).

Reply:
Do we have an example query? It's a bit weird that metadata columns can be propagated through project.

Reply:
This was used in `natural-join.sql`:
```
SELECT nt1.*, nt2.*, nt3.* FROM nt1 natural join nt2 natural join nt3;
```
Which would otherwise fail to expand `nt2.*` to include the key column.

Reply:
Unfortunately, this also breaks the ordering in `udf/postgreSQL/udf-join.sql`.

Reply:
To fix this, we could say that hidden columns are only propagated through `Project` with a `Star` in the `projectList`. Ideally, we'd only propagate the hidden columns that match the Star's qualifier, but I'm not sure if that's possible at that stage.

Reply:
I changed this so hidden output are only propagated if a Project already has hidden output. This allows us to propagate hidden output through nested NATURAL/USING JOINs.